nemo/utils/timers.py

```diff
@@ -0,0 +1,130 @@
+"""
+This module support timing of code blocks.
+"""
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+import numpy as np
+import torch
+
+__all__ = ["NamedTimer"]
+
+
+class NamedTimer(object):
+    """
+    A timer class that supports multiple named timers.
+    A named timer can be used multiple times, in which case the average
+    dt will be returned.
+    A named timer cannot be started if it is already currently running.
+    Use case: measuring execution of multiple code blocks.
+    """
+
+    def __init__(self, mean=True):
+        """
+        mean - default behaviour. If True mean dt is returned, else a list.
+        """
+        self.mean = mean
+        self.reset()
+
+    def __getitem__(self, k):
+        return self.get(k)
+
+    def reset(self, name=None):
+        """
+        Resents all / specific timer
+        """
+        if name is None:
+            self.timers = {}
+        else:
+            self.timers[name] = {}
+
+    def start(self, name=""):
+        """
+        Starts measuring a named timer.
+        """
+        timer_data = self.timers.get(name, {})
+
+        if "start" in timer_data:
+            raise RuntimeError(f"Cannot start timer = '{name}' since it is already active")
+
+        # synchronize pytorch cuda execution if supported
+        if torch.cuda.is_initialized():
+            torch.cuda.synchronize()
+
+        timer_data["start"] = time.time()
+
+        self.timers[name] = timer_data
+
+    def stop(self, name=""):
+        """
+        Stops measuring a named timer.
+        """
+        timer_data = self.timers.get(name, None)
+        if (timer_data is None) or ("start" not in timer_data):
+            raise RuntimeError(f"Cannot end timer = '{name}' since it is not active")
+
+        # synchronize pytorch cuda execution if supported
+        if torch.cuda.is_initialized():
```

Comment:
Same comment as above, please revert to hurt training time.

Reply:
Will be resolved as above.